### PR TITLE
Add GPU implementation for Unique[WithCounts] ops

### PR DESCRIPTION
**Pull request by benbarsdell**
Adds a GPU op kernel supporting the Unique and UniqueWithCounts ops (not the V2 versions).

cc @nluehr

### CODE REVIEW

**Comment by benbarsdell** [comment]
Changed to WIP because I see from the test failures that I made an oversight in the contract of this op (I only ran the main tests myself and they didn't catch it). I will update again once I have a fix for it.

**Comment by benbarsdell** [comment]
I've changed the implementation to produce the output in the correct order now. (I also added an explicit test for that).

Removing "WIP", ready for review.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+    //    the second sort permutation to produce an inverse ID mapping. If
+    //    counts are required, also take the adjacent difference between
+    //    segment_ends indices to produce counts.
+    //      output = input[sorted_unique_input_inds]
+    //      inv_sorted_unique_perm[sorted_unique_perm[i]] = i
+    //      counts = adjacent_difference(segment_ends)
+    // 6) Look up unique IDs via the inverse ID mapping and scatter them using
+    //    the original sort permutation to produce the indices output.
+    //      idx[sorted_input_inds] =
+    //          inv_sorted_unique_perm[sorted_input_unique_ids]
```
Comment (sanjoy) [inline tensorflow/core/kernels/unique_op_gpu.cu.cc:235]: Would it be too cumbersome to interleaved a small but non-trivial running example with the algorithm?  That would make the algorithm much easier to understand at a glance.

File: tensorflow/python/kernel_tests/unique_op_test.py [hunk 2]
```diff
@@ -152,6 +152,24 @@ def testShapeInferenceV2(self):
...
+
+  def testOrderedByAppearance(self):
+    x = np.array([3, 5, 3, 4, 1, 4, 9, 8, 6, 3, 5, 7, 8, 8, 4, 6, 4, 2, 5, 6])
+    true_y = np.array([3, 5, 4, 1, 9, 8, 6, 7, 2])
+    true_idx = np.array(
+        [0, 1, 0, 2, 3, 2, 4, 5, 6, 0, 1, 7, 5, 5, 2, 6, 2, 8, 1, 6])
+    y, idx = array_ops.unique(x)
+    tf_y, tf_idx = self.evaluate([y, idx])
+    self.assertAllEqual(tf_y, true_y)
+    self.assertAllEqual(tf_idx, true_idx)
```
Comment (sanjoy) [inline tensorflow/python/kernel_tests/unique_op_test.py:171]: I suspect these are exercising the existing CPU kernel for `Unique` that works on integers.  If that's true, can you please use floats?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/unique_op_test.py (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/python/kernel_tests/unique_op_test.py:171]: The type registrations are identical for CPU and GPU, and the tests are running on the GPU for me.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/unique_op_gpu.cu.cc]: Done.

File: tensorflow/core/kernels/gpu_prim_helpers.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                         size, out);
+}
+
+}  // namespace detail
+
+// Computes keys_out = sorted(keys_in), and indices_out = argsort(keys_in).
+// If keys_out is not required, it can be set to nullptr.
+// If indices_in is nullptr, the range of input indices [0, size) will be used.
+template <typename Tkey, typename Tindex>
+Status GpuRadixSort(OpKernelContext* context, int size, const Tkey* keys_in,
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/gpu_prim_helpers.h:55]: The context is required to do the temporary allocations, which is a large part of the benefit of the helper functions. They're intended to be helpers specifically for implementing op kernels (they reside in the same directory); the underlying functions can still be used directly if needed in other parts of the code.

File: tensorflow/core/kernels/BUILD [hunk 4]
```diff
@@ -501,6 +501,14 @@ cc_library(
     ]),
 )
 
+cc_library(
+    name = "gpu_prim_helpers",
+    hdrs = ["gpu_prim_helpers.h"],
+    deps = if_cuda_or_rocm([
+        ":gpu_prim_hdrs",
+    ]),
+)
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/BUILD:513]: Done.

File: tensorflow/core/kernels/gpu_prim_helpers.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    Tkey* keys_out,            // Optional
+                    const Tindex* indices_in,  // Optional
+                    Tindex* indices_out, int begin_bit = 0,
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/gpu_prim_helpers.h]: Removed (also changed end_bit to num_bits).

File: tensorflow/core/kernels/gpu_prim_helpers.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    TF_RETURN_IF_ERROR(context->allocate_temp(
+        DataTypeToEnum<Tkey>::value, TensorShape({size}), &tmp_keys_out));
+    keys_out = tmp_keys_out.flat<Tkey>().data();
+  }
+  // Determine temporary device storage requirements.
+  Tensor temp_storage;
+  size_t temp_storage_bytes = 0;
+  const auto& cu_stream = GetGpuStream(context);
+  auto err = gpuprim::DeviceRadixSort::SortPairs(
+      NULL, temp_storage_bytes, keys_in, keys_out, indices_in, indices_out,
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/gpu_prim_helpers.h]: Done.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+      TIndex* sorted_unique_perm_ptr = nullptr;
+      AllocateTemp(context, uniq_size, &sorted_unique_perm,
+                   &sorted_unique_perm_ptr, done);
+      if (!context->status().ok()) return;
+
+      // Sort by input index so that output is in order of appearance.
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          (GpuRadixSort(context, uniq_size,
+                        /*keys_in = */ unique_input_inds_ptr,
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/unique_op_gpu.cu.cc]: Done.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+
+    Tensor sorted_input;
+    T* sorted_input_ptr = nullptr;
+    AllocateTemp(context, input_size, &sorted_input, &sorted_input_ptr, done);
+    if (!context->status().ok()) return;
+
+    const T* input_ptr = input.flat<T>().data();
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        (GpuRadixSort(context, input_size, /*keys_in=*/input_ptr,
```
Comment (sanjoy) [inline tensorflow/core/kernels/unique_op_gpu.cu.cc:268]: Extra parens not needed right?

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    OP_REQUIRES_OK_ASYNC(
+        context,
+        GpuInclusivePrefixSum(context, input_size, segment_indicator_iter,
+                              sorted_input_unique_ids_ptr),
+        done);
+
+    // Copy the last element of sorted_input_unique_ids back to the host to
+    // obtain uniq_size.
+    ScratchSpace<TIndex> last_idx_host(context, 1, /*on_host=*/true);
+    if (input_size > 0) {
```
Comment (sanjoy) [inline tensorflow/core/kernels/unique_op_gpu.cu.cc:294]: Why not early exit if `input_size` is `0`?  (Please add a comment.)

**Review by sanjoy — COMMENTED** [review]
Only minor comments.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc (hunk 8 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/unique_op_gpu.cu.cc]: Removed.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc (hunk 9 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/unique_op_gpu.cu.cc]: Added early exit instead.

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+      Tensor sorted_unique_perm;
+      TIndex* sorted_unique_perm_ptr = nullptr;
+      AllocateTemp(context, uniq_size, &sorted_unique_perm,
+                   &sorted_unique_perm_ptr, done);
+      if (!context->status().ok()) return;
+
+      // Sort by input index so that output is in order of appearance.
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          (GpuRadixSort(context, uniq_size,
```
Comment (sanjoy) [inline tensorflow/core/kernels/unique_op_gpu.cu.cc:366]: Unnecessary parens

File: tensorflow/core/kernels/unique_op_gpu.cu.cc [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+// unique value's range in the sorted input (the last element is always set
+// to input_size).
+template <typename TIndex>
+Status ExtractFirstOccurrenceIndices(const GPUDevice& d, int64 input_size,
+                                     int64 uniq_size,
+                                     const TIndex* sorted_input_inds,
+                                     const TIndex* sorted_input_unique_ids,
+                                     TIndex* unique_input_inds,
+                                     TIndex* segment_ends) {
+  if (input_size == 0) return Status::OK();
```
Comment (sanjoy) [inline tensorflow/core/kernels/unique_op_gpu.cu.cc:89]: This can be a `CHECK` right now that you have an early exit for empty input?  Same elsewhere.

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Review by sanjoy — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by benbarsdell** [comment]
I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to https://github.com/tensorflow/tensorflow/pull/46172). Is this important?

[1] https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops_impl.h#L277-L280

**Comment by sanjoy** [comment]
Hi,

I'm getting failures when I run `tensorflow/python/kernel_tests:unique_op_test_gpu` internally:

```
UniqueWithCountsTest.testInt32
third_party/tensorflow/python/kernel_tests/unique_op_test.py", line 183, in testInt32
    self.assertEqual(len(tf_y), len(np.unique(x)))
AssertionError: 2 != 8
```

(I had to change the `tf_py_test` to a `cuda_py_test` to run the test internally.  I don't think that's necessary in OSS, but you might want to do the same to avoid any confounding issues.)

**Comment by benbarsdell** [comment]
Hm, what GPU is it running on when it fails? Also is it just that test or many?

**Comment by sanjoy** [comment]
> I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to #46172). Is this important?

Yes, that's important, sorry for not catching that.  (The continuation could be run on another thread.)

**Comment by sanjoy** [comment]
> > I just noticed that existing uses of `ThenExecute` all seem to have `ScopedActivateExecutorContext` inside the lambda callback (e.g., [1]), but I didn't add that inside this PR (the same also applies to #46172). Is this important?
> 
> Yes, that's important, sorry for not catching that. (The continuation could be run on another thread.)

CC @kaixih I believe your PR also needs this fix, can you please send a new PR addressing this?  Thanks!

**Comment by sanjoy** [comment]
> Hm, what GPU is it running on when it fails? Also is it just that test or many?

It was running on P100, and TF was compiled with clang.

All of these tests fail:

```
UniqueTest.testInt32
UniqueTest.testInt32OutIdxInt64
UniqueTest.testOrderedByAppearance
UniqueWithCountsTest.testInt32
UniqueWithCountsTest.testInt32OutIdxInt64
UniqueWithCountsTest.testOrderedByAppearance
```

And the failure mode is the same -- the output has only two elements.

E.g.

```
third_party/tensorflow/python/kernel_tests/unique_op_test.py", line 49, in testInt32OutIdxInt64
    self.assertEqual(len(tf_y), len(np.unique(x)))
AssertionError: 2 != 8
```

**Comment by benbarsdell** [comment]
I managed to build using clang-12, sm_60, and CUDA 11.0u1, but still haven't been able to repro the failure.

I don't suppose changing the `bool` on this line to `TIndex` helps? https://github.com/tensorflow/tensorflow/pull/47082/files#diff-39b40cf87839d63bdeeffbef44effbbd5ff5f3549006091b32dbdfa27f8b8ca5R294

**Comment by sanjoy** [comment]
> I managed to build using clang-12, sm_60, and CUDA 11.0u1, but still haven't been able to repro the failure.
> 
> I don't suppose changing the `bool` on this line to `TIndex` helps? https://github.com/tensorflow/tensorflow/pull/47082/files#diff-39b40cf87839d63bdeeffbef44effbbd5ff5f3549006091b32dbdfa27f8b8ca5R294

This link does not point to a line.

**Comment by benbarsdell** [comment]
I think you need to expand the contracted diff of `unique_op_gpu.cu.cc` to see the line.

It's line 294 here:
```
    // Create a fancy input iterator to indicate segment boundaries.
    gpuprim::TransformInputIterator<bool, SegmentIndicatorFunctor<T, TIndex>,
                                    gpuprim::CountingInputIterator<TIndex>>
        segment_indicator_iter(0, {sorted_input_ptr});
```
changing to this:
```
    gpuprim::TransformInputIterator<TIndex, SegmentIndicatorFunctor<T, TIndex>,
```

**Comment by sanjoy** [comment]
> I think you need to expand the contracted diff of `unique_op_gpu.cu.cc` to see the line.
> 
> It's line 294 here:
> 
> ```
>     // Create a fancy input iterator to indicate segment boundaries.
>     gpuprim::TransformInputIterator<bool, SegmentIndicatorFunctor<T, TIndex>,
>                                     gpuprim::CountingInputIterator<TIndex>>
>         segment_indicator_iter(0, {sorted_input_ptr});
> ```
> 
> changing to this:
> 
> ```
>     gpuprim::TransformInputIterator<TIndex, SegmentIndicatorFunctor<T, TIndex>,
> ```

Wow, that seems to work.  Trying to check in now.

**Comment by benbarsdell** [comment]
I see this was merged, but then reverted in [this nondescript commit](https://github.com/tensorflow/tensorflow/commit/b6a25c7641ac37426b1dadd21b74f814c36f1738)?

I've pushed the `bool` fix to this branch so that it's complete; but now it says there is a merge conflict? I'm a bit confused.

**Comment by sanjoy** [comment]
Ben, this was auto-reverted because `//tensorflow/python/debug:analyzer_cli_test_gpu` failed.  The failure seems trivial, I'll try to reland with a fix.

**Comment by sanjoy** [comment]
Went in as https://github.com/tensorflow/tensorflow/commit/02585ace8ca787613637713aebfd581a6dc2a12a.

**Pull request closed without merging**